File: 072 - Counting Fractions/Haskell/pe_072.hs
{-
file:   pe_072.hs
title:  Counting fractions
author: Alexander Gosselin
e-mail: alexandergosselin@gmail.com
        alexander.gosselin@alumni.ubc.ca
date:   February 7, 2016

link:   <https://projecteuler.net/problem=5>

note:   Regretfully, this is a copy of Zach Denton's solution.
        <https://zach.se/project-euler-solutions/72/>
        I had already solved the problem in C++, but I was having a
        hard time writing a memory-efficient solution in Haskell. My
        attempt at a solution is commented out below. It can only be
        used to compute totientSum n in reasonable time for n less than 
        a few tens of thousands.
        
references:
    Farey sequence:
        <http://mathworld.wolfram.com/FareySequence.html>
        <https://en.wikipedia.org/wiki/Farey_sequence>
-}

farey :: Int -> Int
farey = (map farey' [0..] !!)
  where
    farey' n = (quot (n*(n + 3)) 2) 
               - sum [farey $ quot n x | x <- [2..n]]

totientSum n = farey n - farey 1

pe_072 = totientSum 1000000 

main :: IO ()
main = do
  print pe_072

{-
-- <https://www.reddit.com/2tpqip/>
--applyEvery :: (a -> a) -> Int -> [a] -> [a]
--applyEvery f n = zipWith ($) (cycle (replicate (n - 1) id ++ [f]))

-- more memory efficient than above implementation
applyEvery :: (a -> a) -> Int -> [a] -> [a]
applyEvery f n xs = xf ++ (\(y:ys) -> f y : applyEvery f n ys) xb
  where
    (xf, xb) = splitAt (n - 1) xs

totients :: [Int]
totients = 1 : sieve [2..] [2..]
  where 
    sieve (x:xs) (y:ys) 
        | x == y    = (y - 1) : sieve xs (propagatePrime x ys)  
        | otherwise = y : sieve xs ys
    propagatePrime j = applyEvery (\x -> (quot x j)*(j - 1))

totientSum :: Int -> Int
totientSum n = sum $ take n totients
-}

File: 072 - Counting Fractions/C++/pe_072.cc
/*
file:   pe_072.hs
title:  Counting fractions
author: Alexander Gosselin
e-mail: alexandergosselin@gmail.com
        alexander.gosselin@alumni.ubc.ca
date:   February 8, 2016

link:   <https://projecteuler.net/problem=72>

copyright:  (C) 2016 Alexander Gosselin
license:    GNU General Public License v3.0
            <http://www.gnu.org/licenses/gpl-3.0.en.html>
*/

#include <iostream>
#include <vector>
using namespace std;

static const int N = 1000000;

int main() {
  unsigned long int totient_sum = 0;
  vector<int> sieve; // sieve generating the totients
  for (int i = 0; i <= N; sieve.push_back(i++));
  for (int i = 2; i <= N; i++) {
    if (i == sieve[i]) { // i is prime
      sieve[i] -= 1;
      for (int j = 2*i; j <= N; j += i) {
        sieve[j] /= i;
        sieve[j] *= i - 1;
      }
    }
    totient_sum += sieve[i];
  }
  cout << totient_sum << endl;
  return 0;
}

